Synthetic source index out of bounds exception

As part of running a elastic/logging benchmark with synthetic source enabled, I noticed the following stacktrace in the Elasticsearch logs:

```
Caused by: org.elasticsearch.common.io.stream.NotSerializableExceptionWrapper: index_out_of_bounds_exception: null
        at org.apache.lucene.codecs.lucene90.Lucene90DocValuesProducer$TermsDict.seekExact(Lucene90DocValuesProducer.java:1114) ~[lucene-core-9.8.0.jar:?]
        at org.apache.lucene.codecs.lucene90.Lucene90DocValuesProducer$BaseSortedDocValues.lookupOrd(Lucene90DocValuesProducer.java:982) ~[lucene-core-9.8.0.jar:?]
        at org.apache.lucene.index.SingletonSortedSetDocValues.lookupOrd(SingletonSortedSetDocValues.java:95) ~[lucene-core-9.8.0.jar:?]
        at org.elasticsearch.index.mapper.SortedSetDocValuesSyntheticFieldLoader$ImmediateDocValuesLoader.write(SortedSetDocValuesSyntheticFieldLoader.java:198) ~[elasticsea
rch-8.12.0-SNAPSHOT.jar:?]
        at org.elasticsearch.index.mapper.SortedSetDocValuesSyntheticFieldLoader.write(SortedSetDocValuesSyntheticFieldLoader.java:125) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?
]
        at org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader.write(ObjectMapper.java:711) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader.write(ObjectMapper.java:711) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.source(SourceLoader.java:126) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.elasticsearch.search.lookup.SyntheticSourceProvider$SyntheticSourceLeafLoader.getSource(SyntheticSourceProvider.java:74) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?
]
        at org.elasticsearch.search.lookup.SyntheticSourceProvider.getSource(SyntheticSourceProvider.java:40) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.elasticsearch.search.lookup.LeafSearchLookup.lambda$new$0(LeafSearchLookup.java:39) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.elasticsearch.script.AbstractFieldScript.extractFromSource(AbstractFieldScript.java:106) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.elasticsearch.script.AbstractFieldScript.emitFromSource(AbstractFieldScript.java:126) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.elasticsearch.script.LongFieldScript$1$1.execute(LongFieldScript.java:28) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.elasticsearch.script.AbstractFieldScript.runForDoc(AbstractFieldScript.java:158) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.elasticsearch.search.runtime.AbstractLongScriptFieldQuery.matches(AbstractLongScriptFieldQuery.java:33) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.elasticsearch.search.runtime.AbstractLongScriptFieldQuery.matches(AbstractLongScriptFieldQuery.java:21) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.elasticsearch.search.runtime.AbstractScriptFieldQuery$1$1.matches(AbstractScriptFieldQuery.java:77) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.apache.lucene.search.Weight$DefaultBulkScorer.scoreRange(Weight.java:295) ~[lucene-core-9.8.0.jar:?]
        at org.apache.lucene.search.Weight$DefaultBulkScorer.score(Weight.java:236) ~[lucene-core-9.8.0.jar:?]
        at org.elasticsearch.search.internal.CancellableBulkScorer.score(CancellableBulkScorer.java:45) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.apache.lucene.search.BulkScorer.score(BulkScorer.java:38) ~[lucene-core-9.8.0.jar:?]
        at org.elasticsearch.search.internal.ContextIndexSearcher.searchLeaf(ContextIndexSearcher.java:526) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:455) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.elasticsearch.search.internal.ContextIndexSearcher.lambda$search$4(ContextIndexSearcher.java:370) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
        at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) ~[elasticsearch-8.12.0-SNAPSHOT.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.lang.Thread.run(Thread.java:1583) ~[?:?]
```

I don't have more information yet. I will update this issue when I know more.

--- a/modules/lang-painless/src/internalClusterTest/java/org/elasticsearch/painless/search/SyntheticSourceIT.java
+++ b/modules/lang-painless/src/internalClusterTest/java/org/elasticsearch/painless/search/SyntheticSourceIT.java
@@ -10,7 +10,6 @@ package org.elasticsearch.painless.search;
 
 import org.elasticsearch.action.index.IndexRequestBuilder;
 import org.elasticsearch.action.support.WriteRequest;
-import org.elasticsearch.core.CheckedFunction;
 import org.elasticsearch.index.query.QueryBuilders;
 import org.elasticsearch.painless.PainlessPlugin;
 import org.elasticsearch.plugins.Plugin;
@@ -31,12 +30,12 @@ public class SyntheticSourceIT extends ESIntegTestCase {
         return Collections.singleton(PainlessPlugin.class);
     }
 
-    public void testText() throws Exception {
-        createIndex(b -> b.field("type", "text").field("store", true));
+    public void testSearchUsingRuntimeField() throws Exception {
+        createIndex();
 
         int numDocs = between(1000, 5000);
         for (int i = 0; i < numDocs; i++) {
-            IndexRequestBuilder indexRequest = client().prepareIndex("test").setSource("id", "" + i, "field", "n" + i);
+            IndexRequestBuilder indexRequest = client().prepareIndex("test").setSource("id", "" + i);
             if (randomInt(100) < 5) {
                 indexRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
             }
@@ -46,7 +45,7 @@ public class SyntheticSourceIT extends ESIntegTestCase {
         assertNoFailures(client().prepareSearch("test").setQuery(QueryBuilders.rangeQuery("long_id").from(0)));
     }
 
-    private void createIndex(CheckedFunction<XContentBuilder, XContentBuilder, IOException> fieldMapping) throws IOException {
+    private void createIndex() throws IOException {
         XContentBuilder mapping = JsonXContent.contentBuilder();
         mapping.startObject();
         {
@@ -63,9 +62,6 @@ public class SyntheticSourceIT extends ESIntegTestCase {
             mapping.endObject();
             mapping.startObject("properties");
             mapping.startObject("id").field("type", "keyword").endObject();
-            mapping.startObject("field");
-            fieldMapping.apply(mapping);
-            mapping.endObject();
             mapping.endObject();
         }
         mapping.endObject();

--- a/modules/lang-painless/src/internalClusterTest/java/org/elasticsearch/painless/search/SyntheticSourceIT.java
+++ b/modules/lang-painless/src/internalClusterTest/java/org/elasticsearch/painless/search/SyntheticSourceIT.java
@@ -1,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.painless.search;
+
+import org.elasticsearch.action.index.IndexRequestBuilder;
+import org.elasticsearch.action.support.WriteRequest;
+import org.elasticsearch.core.CheckedFunction;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.painless.PainlessPlugin;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.test.ESIntegTestCase;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.json.JsonXContent;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertNoFailures;
+
+public class SyntheticSourceIT extends ESIntegTestCase {
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return Collections.singleton(PainlessPlugin.class);
+    }
+
+    public void testText() throws Exception {
+        createIndex(b -> b.field("type", "text").field("store", true));
+
+        int numDocs = between(1000, 5000);
+        for (int i = 0; i < numDocs; i++) {
+            IndexRequestBuilder indexRequest = client().prepareIndex("test").setSource("id", "" + i, "field", "n" + i);
+            if (randomInt(100) < 5) {
+                indexRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+            }
+            indexRequest.get();
+        }
+        client().admin().indices().prepareRefresh("test").get();
+        assertNoFailures(client().prepareSearch("test").setQuery(QueryBuilders.rangeQuery("long_id").from(0)));
+    }
+
+    private void createIndex(CheckedFunction<XContentBuilder, XContentBuilder, IOException> fieldMapping) throws IOException {
+        XContentBuilder mapping = JsonXContent.contentBuilder();
+        mapping.startObject();
+        {
+            mapping.startObject("_source");
+            mapping.field("mode", "synthetic");
+            mapping.endObject();
+        }
+        {
+            mapping.startObject("runtime");
+            mapping.startObject("long_id");
+            mapping.field("type", "long");
+            mapping.field("script", "emit(Long.parseLong(params._source.id));");
+            mapping.endObject();
+            mapping.endObject();
+            mapping.startObject("properties");
+            mapping.startObject("id").field("type", "keyword").endObject();
+            mapping.startObject("field");
+            fieldMapping.apply(mapping);
+            mapping.endObject();
+            mapping.endObject();
+        }
+        mapping.endObject();
+
+        assertAcked(client().admin().indices().prepareCreate("test").setMapping(mapping).get());
+    }
+}

--- a/server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java
+++ b/server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java
@@ -32,6 +32,7 @@ import org.elasticsearch.index.engine.Engine;
 import org.elasticsearch.index.fielddata.FieldDataContext;
 import org.elasticsearch.index.fielddata.IndexFieldData;
 import org.elasticsearch.index.fielddata.IndexOrdinalsFieldData;
+import org.elasticsearch.index.mapper.DocumentMapper;
 import org.elasticsearch.index.mapper.IdLoader;
 import org.elasticsearch.index.mapper.MappedFieldType;
 import org.elasticsearch.index.mapper.NestedLookup;
@@ -167,7 +168,7 @@ final class DefaultSearchContext extends SearchContext {
 
             Engine.Searcher engineSearcher = readerContext.acquireSearcher("search");
             int maximumNumberOfSlices;
-            if (indexService.mapperService().documentMapper().sourceMapper().isSynthetic()) {
+            if (hasSyntheticSource(indexService)) {
                 // accessing synthetic source is not thread safe
                 maximumNumberOfSlices = 1;
             } else {
@@ -219,6 +220,14 @@ final class DefaultSearchContext extends SearchContext {
                 close();
             }
         }
+    }
+
+    private static boolean hasSyntheticSource(IndexService indexService) {
+        DocumentMapper documentMapper = indexService.mapperService().documentMapper();
+        if (documentMapper != null) {
+            return documentMapper.sourceMapper().isSynthetic();
+        }
+        return false;
     }
 
     static long getFieldCardinality(String field, IndexService indexService, DirectoryReader directoryReader) {

--- a/server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java
+++ b/server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java
@@ -166,13 +166,20 @@ final class DefaultSearchContext extends SearchContext {
             this.indexShard = readerContext.indexShard();
 
             Engine.Searcher engineSearcher = readerContext.acquireSearcher("search");
-            int maximumNumberOfSlices = determineMaximumNumberOfSlices(
-                executor,
-                request,
-                resultsType,
-                enableQueryPhaseParallelCollection,
-                field -> getFieldCardinality(field, readerContext.indexService(), engineSearcher.getDirectoryReader())
-            );
+            int maximumNumberOfSlices;
+            if (indexService.mapperService().documentMapper().sourceMapper().isSynthetic()) {
+                // accessing synthetic source is not thread safe
+                maximumNumberOfSlices = 1;
+            } else {
+                maximumNumberOfSlices = determineMaximumNumberOfSlices(
+                    executor,
+                    request,
+                    resultsType,
+                    enableQueryPhaseParallelCollection,
+                    field -> getFieldCardinality(field, readerContext.indexService(), engineSearcher.getDirectoryReader())
+                );
+
+            }
             if (executor == null) {
                 this.searcher = new ContextIndexSearcher(
                     engineSearcher.getIndexReader(),

--- a/server/src/test/java/org/elasticsearch/search/DefaultSearchContextTests.java
+++ b/server/src/test/java/org/elasticsearch/search/DefaultSearchContextTests.java
@@ -395,6 +395,8 @@ public class DefaultSearchContextTests extends ESTestCase {
         when(indexShard.getThreadPool()).thenReturn(threadPool);
 
         IndexService indexService = mock(IndexService.class);
+        MapperService mapperService = mock(MapperService.class);
+        when(indexService.mapperService()).thenReturn(mapperService);
 
         try (Directory dir = newDirectory(); RandomIndexWriter w = new RandomIndexWriter(random(), dir)) {
 

